File: cpp/allergies/allergies.cpp
#include "allergies.h"
#include <array>

namespace allergies {
static const std::array<std::string, 8> allergens = {
    "eggs",     "peanuts",   "shellfish", "strawberries",
    "tomatoes", "chocolate", "pollen",    "cats",
};

allergy_test::allergy_test(int score) {
  for (size_t i = 0; i < allergens.size(); i++) {
    if (score & (1 << i)) {
      allergies.emplace(allergens[i]);
    }
  }
}

bool allergy_test::is_allergic_to(const std::string &allergen) {
  return allergies.count(allergen) != 0;
}

const std::unordered_set<std::string> &allergy_test::get_allergies() {
  return allergies;
}
} // namespace allergies

File: lua/allergies/allergies.lua
local allergens = {
	"eggs",
	"peanuts",
	"shellfish",
	"strawberries",
	"tomatoes",
	"chocolate",
	"pollen",
	"cats",
}

local function list(score)
	local result = {}

	for i, allergen in ipairs(allergens) do
		if score & (2 ^ (i - 1)) ~= 0 then
			table.insert(result, allergen)
		end
	end

	return result
end

local function allergic_to(score, which)
	for i, allergen in ipairs(allergens) do
		if allergen == which and score & (2 ^ (i - 1)) ~= 0 then
			return true
		end
	end

	return false
end

return {
	list = list,
	allergic_to = allergic_to,
}
